bridge/sam/pool.go

```diff
@@ -0,0 +1,377 @@
+	defer p.changeValidatorsLock.Unlock()
+
+	oldValidators := p.validators
+	oldThreshold := p.signatureThreshold //nolint
+
+	p.validators = validators
+	atomic.StoreUint64(&p.signatureThreshold, threshold)
+
+	var maybeDemotableHashes []types.Hash
+	if removed := common.DiffAddresses(oldValidators, validators); len(removed) > 0 {
+		maybeDemotableHashes = p.messageSignatures.RemoveSignatures(removed)
+	}
+
+	if oldThreshold != threshold {
+		// we need to check all messages if threshold changes
+		p.tryToPromoteAndDemoteAll()
+	} else if len(maybeDemotableHashes) > 0 {
+		for _, hash := range maybeDemotableHashes {
+			p.tryToDemote(hash)
+		}
+	}
+}
+
+// canPromote return the flag indicating it's possible to change status to ready
+// message need to have enough signatures and be known by pool for promotion
+func (p *pool) canPromote(hash types.Hash) bool {
+	isKnown := p.isMessageKnown(hash)
+	numSignatures := p.messageSignatures.GetSignatureCount(hash)
+	threshold := atomic.LoadUint64(&p.signatureThreshold)
+
+	return isKnown && numSignatures >= threshold
+}
+
+// canDemote return the flag indicating it's possible to change status to pending
+func (p *pool) canDemote(hash types.Hash) bool {
+	return !p.canPromote(hash)
+}
+
+// tryToPromote checks the number of signatures and threshold and update message status to ready if need
+func (p *pool) tryToPromote(hash types.Hash) {
+	if p.canPromote(hash) {
+		p.promote(hash)
+	}
+}
+
+// tryToDemote checks the number of signatures and threshold and update message status to pending if need
+func (p *pool) tryToDemote(hash types.Hash) {
+	if p.canDemote(hash) {
+		p.demote(hash)
+	}
+}
+
+// tryToPromoteAndDemoteAll iterates all messages and update its statuses
+func (p *pool) tryToPromoteAndDemoteAll() {
+	threshold := atomic.LoadUint64(&p.signatureThreshold)
+
+	p.messageSignatures.RangeMessages(func(entry *signedMessageEntry) bool {
+		hash := entry.Hash
+		isKnown := p.isMessageKnown(hash)
+		numSignatures := entry.NumSignatures()
+
+		if numSignatures >= threshold && isKnown {
+			p.promote(hash)
+		} else {
+			p.demote(hash)
+		}
+
+		return true
+	})
+}
+
+// promote change message status to ready
+func (p *pool) promote(hash types.Hash) {
+	p.readyMap.Store(hash, true)
+}
+
+// promote change message status to pending
+// it deletes instead of unsetting for less-complexity on getting ready messages
+func (p *pool) demote(hash types.Hash) {
+	p.readyMap.Delete(hash)
+}
+
+func (p *pool) setMessageBody(hash types.Hash, body []byte) {
+	p.messageMap.Store(hash, body)
+}
+
+func (p *pool) getMessageBody(hash types.Hash) []byte {
+	raw, existed := p.messageMap.Load(hash)
+	if !existed {
+		return nil
+	}
+
+	body, ok := raw.([]byte)
+	if !ok {
+		// should not reach
+
+		return nil
+	}
+
+	return body
+}
+
+// signedMessageEntry is representing the data stored in messageSignaturesStore
+type signedMessageEntry struct {
+	Hash           types.Hash
+	Signatures     sync.Map
+	SignatureCount int64
```

Comment:
Perhaps this field can be of type `uint64` - this would remove the `uint64()` casting of the return values in `NumSignatures`, `IncrementNumSignatures` and `DecrementNumSignatures`